### ISSUE TITLE
Proposal: Add C++ code formatting

### ISSUE THREAD

**Issue by bcumming** [arbor-sim/arbor#1328]
**Proposal**

Add clang-format rules for the C++ code in Arbor.

**Motivation**

To ensure consistent, unsurprising formatting of code in the code base. Also to make it easier for developers who don't want to worry about formatting.

**Decisions**

There are some decisions to make

- Is the use of the tool required to solve formatting problems in the code today?
- Which tool to use (clang-format appears to be the preferred solution if this is implemented)
- What are the formatting rules?
  - can we actually implement our rules using the chosen linter?
- Should the formatter be automated?
  - if so, what scope. e.g. as a script for developers to run on their commits, or further as a test in CI.
- Should use of the formatter be mandatory?
  - this would require an automated pass/fail test in GitHub Actions.

Depending on these decisions, we might have a range of solutions ranging from a simple script that developers can optionally run on their commit, to automated testing that will reject a PR whose changes don't satisfy the linting rules.

*Note*: the discussion here is specifically about code formatting for the C++ library. Using tools like black for the Python code is a separate topic.

**Comment by brenthuisman** [comment]
Here Nora's attempt at capturing current style: https://gist.github.com/noraabiakar/4335b472b1e0cc48c3b6e15668d7f9e9

**Comment by bcumming** [comment]
My personal opinion on this is that having a set of clang-format rules that can be optionally used by a developer might be a good thing, however I am against mandatory enforcement or automation in CI processes woul.

To ensure that PRs are clear and easy to review it should be a hard requirement that the formatter should only be applied to code that is changed or added in the PR, and not to unrelated code. As a corollary to this, we could enforce that PRs should not contain many unrelated changes.

I am not fond of mandatory use of the tool for the following reasons.

*existing code formatting is good enough* the Arbor code is quite consistently formatted. There is some small variation between the developers, however we all colour between the lines.
- there are very few formatting comments in our PR reviews, and when they are they don't lead to arguments.

*clang-format can't reproduce our existing norms*: @bcumming , @halfflat and @noraabiakar  have experimented more than once with trying to get output from clang-format that was "good enough", and were never satisfied with the results.
- clang format does not appear to update their rules very often
- when laying out mathematical expressions, or complex code, we often use indenting and spacing to make the code logical, which clang-format won't respect. This will lead to demonstrably worse code formatting where it is needed most.
- clang-format doesn't have common sense: we have a set of rulees, e.g. "always use `{}` and a new line for if statements that contain a single statement. This rule is followed 99% of the time, but we all break it in extreme cases where it makes the code harder to follow (e.g. when we have to do lots of C-style error checking). This isn't a great example, because it clang-format can't test for or assert this formatting requirement!

*Automation in CI*: if the formatting is mandatory, there have to be automated tests, which would probably best be performed as a GitHub action. This is actually non-trivial to implement in a robust way that is easy for users.
- I have seen developers in my team waste days because the behavior of clang-format version running on the OS + VM + phase of the moon on the CI server expects formatting that can't be reproduced locally on the developer's machines.
- Writing a script that will enforce formatting only on code that is changed *relative to the master branch* would be difficult for long lived branches with many commits that are to be squashed. This might force changes to the developer workflow.

*Technical debt and workflow complexity*: it is a myth that automated code formatting in C++ code is "free", and there are overheads of implementing it:
- automated testing and robust scripts for performing formatting and checking will require constant maintenance in the formatting scripts and the CI configurations.
- enforcing the use of the formatting tool will complicate the development and PR process.

*In summary*, it is my position that the Arbor code does not have any pressing problems with how formatting is performed or enforced in our code that require taking on the costs of enforced code testing.
- increased testing complexity
- increased developer workflow complexity
- lower quality formatting

**Comment by thorstenhater** [comment]
Hi,

I brought this up a few weeks ago since some of the rules in the style guide are unintuitive to me and I wanted
an automated solution to fix those things. There is more reasons for doing this, mainly onboarding of new people
and especially one-time contributors. This is thinking into future, where more people might want to add something
to arbor.

In the debate on the actual rules I am impartial, I value consistency and automation over aesthetics in this case.
I agree on arbor's formatting being ok, although a bit unusual in places (which is why automation helps consistency).
My best example is spacing around operators here, I have rarely seen `i==j` instead of `i == j`, which should not
imply that being bad as such.

As for the implementation I would not suggest enforcing this rule on developers by rejecting their PR based on
formatting, but rather by taking it out of their hands by means of a commit hook and/or regular automatted formatting
of the whole project.

After test-driving the style file for two PRs and maybe two weeks I have to agree that it hardly reproduces our guide due
to clang-format's rigidity. However, as said above, I would be willing to comprise on the style itself if that means automated
and consistent results. That might be in part due to the fact that I would change some things in it anyway, to be honest.

Summary
=======

My position in a nutshell
- I think it helps one-time contributors and newcomers of which we hope to get more.
- Consistency and automation more important than perfection.

**Comment by halfflat** [comment]
I feel the problems solved by an automatic formatter can be solved by other means, entailing a different set of compromises.

Why have formatting guidelines at all?
* To add clarity, code comprehension, and general ease of reading for developers.
* To reduce the chance of inadvertent bugs arising from misleading layout.

These are 'soft' goals, in that they are hard to define in terms of formal rules, and different guidelines will be better or worse rather than correct or incorrect.

In my view, consistency plays a very important role in the aesthetics of code, which in turn adds comprehension and reduces irritation. But it should be suborned to these goals, not replace them.

There are only a few formatting tools that work well with C++, and clang-format appears to be the most robust. But it does not admit nuance, and will screw up expression formatting inasmuch as it has no accounting for spacing helping to delineate logical boundaries or differences in precedences. If we were to adopt clang-format as part of the development process, we will be adding a burden: either to work around its efforts where they make the code worse, or to accept its decisions and deal with code that is less satisfactory. This stands even if the cost of incorporating it into the process were negligible.

Code formatting does pose an additional barrier to contributions from those unfamiliar with the conventions in the code base. This, I believe, can be ameliorated by:
* Reviewing, updating and condensing our existing guidelines, and making them prominent.
* Possibly separating formatting from other review in PRs: if a contribution, while legible, deviates from the guidelines, we can accept it and then ourselves correct the formatting before merging, as part of the editorial process. Here 'ourselves' stands for anyone responsible for PR review and merge.

Clang-format does allow the possibility of updates to its code for new style settings, but (according to their documentation) only for well established projects with a well-defined set of code formatting guidelines and dozens of contributors. If we can survive without automatic formatting up to that point, perhaps then we could revisit clang-format as an option, with a view to adapting it to our conventions.

**Comment by noraabiakar** [comment]
I find clang-format, and other C++ formatting tools to be quite inflexible. I'm not particularly enthusiastic for their inclusion, but I'm willing to compromise.

Our code base, according to clang-format, is inconsistent across most dimensions, but I would argue that our code is readable and consistent where it matters and doesn't need reformatting. Most of our style and format preferences can be easily inferred from reading the code or the coding-guidelines, so it's not a huge burden for new contributors to figure it out. And we rarely have issues with formatting when reviewing PRs, which suggests that we don't need to change our process. For those reasons, I don't think clang-format is necessary at this stage. 

On the other hand, I like the idea of fully consistent code. For our code to be fully consistent, we'd have to reformat the whole code base, and I'm not certain the results will be favorable. If we don't reformat our code base and only enforce the style for new contributions, our code won't be fully consistent. So consistency seems to come at the price of partially rewriting our code base, and I don't think that's a good idea.

**Comment by thorstenhater** [comment]
Note: this ties into #1317

**Comment by halfflat** [comment]
Next step: @halfflat (me) will get interested parties together to go through the current coding guidelines, update them, renove 'todo's etc.

**Comment by thorstenhater** [comment]
https://optimal-codestyle.github.io/ has some interesting points on visual perception and formatting. Some of which I do not like stylistically, but make sense.

**Comment by halfflat** [comment]
The placement of a space after an identifier used as a function (or function-like expression) just seems crazy to me: now it doesn't look like a function call anymore. Maybe this is something specific to having a maths background?

For me, something like this is sane:
```
    auto result = frobnicate(x);
    if (result) return x;
```
but the following really is not:
```
    auto result = frobnicate (x);    // doesn't look like a function, but is.
    if(result) return(x);    // both 'if' and 'return' look like functions, but aren't.
```

**Comment by thorstenhater** [comment]
This has gone stale.
